File: src/main/gen.ts
/** AI element generation — turn a text description into a usable graphic.
 *
 *  The renderer never talks to an image provider directly. All requests go
 *  through this service in the main process, where API keys live (env vars or
 *  `userData/gen-config.json`, written only via the `gen:setApiKey` IPC).
 *
 *  Providers (switchable, persisted in gen-config.json):
 *    - `pollinations` (default) — free, no API key. GET image.pollinations.ai
 *      with the prompt; returns image bytes directly. Verified live.
 *    - `bfl` — Black Forest Labs FLUX (paid). POST /v1/flux-pro-1.1 with a
 *      Bearer key, poll /v1/get_result until Ready, download the sample URL.
 *
 *  Generated images are saved as PNG assets into the `assets` table, so they
 *  appear in the editor's "your graphics" library automatically and can be
 *  dropped onto any page like an imported graphic.
 */
import { mkdirSync, readFileSync, writeFileSync } from "fs";
import { join } from "path";
import type { DB } from "./db";
import { now } from "./db";

const POLLINATIONS_ENDPOINT = "https://image.pollinations.ai/prompt";
const BFL_ENDPOINT = "https://api.bfl.ai/v1/flux-pro-1.1";
const BFL_RESULT_ENDPOINT = "https://api.bfl.ai/v1/get_result";
const MAX_POLL_SECONDS = 120;

export type GenProviderId = "pollinations" | "bfl";

const PROVIDER_IDS: GenProviderId[] = ["pollinations", "bfl"];

export interface GenResult {
  ok: boolean;
  asset?: { id: string; name: string; kind: "png"; dataUri: string };
  error?: string;
}

interface GenConfig {
  provider?: GenProviderId;
  bflApiKey?: string;
}

/** Build the pollinations GET URL (pure — unit tested). */
export function pollinationsUrl(prompt: string, opts?: { width?: number; height?: number; seed?: number }): string {
  const w = opts?.width ?? 768;
  const h = opts?.height ?? 768;
  const seed = opts?.seed ?? Math.floor(Math.random() * 1_000_000);
  const params = new URLSearchParams({
    width: String(w),
    height: String(h),
    nologo: "true",
    seed: String(seed),
  });
  return `${POLLINATIONS_ENDPOINT}/${encodeURIComponent(prompt)}?${params.toString()}`;
}

/** Poll a BFL job until Ready and return the sample URL (pure — unit tested). */
export async function pollBflResult(
  id: string,
  apiKey: string,
  fetchImpl: typeof fetch = fetch,
  maxMs = MAX_POLL_SECONDS * 1000,
): Promise<string> {
  const started = Date.now();
  for (;;) {
    const res = await fetchImpl(`${BFL_RESULT_ENDPOINT}?id=${encodeURIComponent(id)}`, {
      headers: { Authorization: `Bearer ${apiKey}`, Accept: "application/json" },
    });
    if (!res.ok) throw new Error(`FLUX status check failed (${res.status})`);
    const data = (await res.json()) as { status?: string; result?: { sample?: string }; error?: string };
    if (data.error) throw new Error(String(data.error));
    if (data.status === "Ready" && data.result?.sample) return data.result.sample;
    if (Date.now() - started > maxMs) throw new Error("FLUX generation timed out");
    await new Promise((r) => setTimeout(r, 1500));
  }
}

/** Sniff a buffer's mime type from its magic bytes. */
export function sniffImageType(buf: Buffer): "png" | "jpeg" | "webp" {
  if (buf.length > 8 && buf[0] === 0x89 && buf[1] === 0x50 && buf[2] === 0x4e && buf[3] === 0x47) return "png";
  if (buf.length > 3 && buf[0] === 0xff && buf[1] === 0xd8 && buf[2] === 0xff) return "jpeg";
  if (buf.length > 12 && buf.subarray(0, 4).toString("latin1") === "RIFF" && buf.subarray(8, 12).toString("latin1") === "WEBP") {
    return "webp";
  }
  return "png";
}

export class GenService {
  private readonly db: DB;
  private readonly cacheDir: string;
  private readonly configPath: string;

  constructor(opts: { db: DB; cacheDir: string; dataDir: string }) {
    this.db = opts.db;
    this.cacheDir = join(opts.cacheDir, "gen");
    this.configPath = join(opts.dataDir, "gen-config.json");
    mkdirSync(this.cacheDir, { recursive: true });
  }

  provider(): GenProviderId {
    const cfg = this.readConfig();
    return cfg.provider === "bfl" ? "bfl" : "pollinations";
  }

  setProvider(p: string): boolean {
    if (!PROVIDER_IDS.includes(p as GenProviderId)) return false;
    const cfg = this.readConfig();
    cfg.provider = p as GenProviderId;
    this.writeConfig(cfg);
    return true;
  }

  setApiKey(provider: string, key: string): boolean {
    if (provider !== "bfl") return false;
    const cfg = this.readConfig();
    cfg.bflApiKey = key.trim();
    cfg.provider = "bfl";
    this.writeConfig(cfg);
    return true;
  }

  /** True when the active provider can generate right now. */
  configured(): boolean {
    if (this.provider() === "pollinations") return true; // free, no key
    return !!this.readConfig().bflApiKey;
  }

  private readConfig(): GenConfig {
    try {
      return JSON.parse(readFileSync(this.configPath, "utf8")) as GenConfig;
    } catch {
      return {};
    }
  }

  private writeConfig(cfg: GenConfig): void {
    try {
      writeFileSync(this.configPath, JSON.stringify(cfg), { mode: 0o600 });
    } catch {
      /* non-fatal */
    }
  }

  /** Generate an image from a description and save it into the assets library. */
  async generate(prompt: string, opts?: { width?: number; height?: number }): Promise<GenResult> {
    const clean = prompt.trim();
    if (!clean) return { ok: false, error: "Enter a description first." };

    const provider = this.provider();
    let buf: Buffer | null = null;
    try {
      if (provider === "pollinations") {
        const res = await fetch(pollinationsUrl(clean, opts));
        if (!res.ok) throw new Error(`Pollinations returned ${res.status}`);
        buf = Buffer.from(await res.arrayBuffer());
      } else {
        const key = this.readConfig().bflApiKey;
        if (!key) return { ok: false, error: "Add a Black Forest Labs API key to use FLUX." };
        const jobRes = await fetch(BFL_ENDPOINT, {
          method: "POST",
          headers: { Authorization: `Bearer ${key}`, "Content-Type": "application/json" },
          body: JSON.stringify({
            prompt: clean,
            width: opts?.width ?? 1024,
            height: opts?.height ?? 1024,
          }),
        });
        if (!jobRes.ok) throw new Error(`FLUX request failed (${jobRes.status})`);
        const job = (await jobRes.json()) as { id?: string };
        if (!job.id) throw new Error("FLUX returned no job id");
        const sample = await pollBflResult(job.id, key);
        const imgRes = await fetch(sample);
        if (!imgRes.ok) throw new Error(`FLUX image download failed (${imgRes.status})`);
        buf = Buffer.from(await imgRes.arrayBuffer());
      }
    } catch (e) {
      return { ok: false, error: e instanceof Error ? e.message : String(e) };
    }
    if (!buf || buf.length === 0) return { ok: false, error: "Empty image received from provider." };

    // Normalize everything to PNG so the renderer and export treat it uniformly.
    let png = buf;
    if (sniffImageType(buf) !== "png") {
      try {
        const sharp = (await import("sharp")).default;
        png = await sharp(buf).png().toBuffer();
      } catch {
        /* keep original bytes — sharp may be unavailable in tests */
      }
    }

    const id = `gen-${Date.now()}-${Math.floor(Math.random() * 1e6)}`;
    const localPath = join(this.cacheDir, `${id}.png`);
    writeFileSync(localPath, png);
    const dataUri = `data:image/png;base64,${png.toString("base64")}`;

    const name = clean.slice(0, 60);
    this.db.prepare("INSERT INTO assets (id, name, kind, data, created_at) VALUES (?, ?, ?, ?, ?)").run(
      id,
      name,
      "png",
      dataUri,
      now(),
    );

    return { ok: true, asset: { id, name, kind: "png", dataUri } };
  }
}

File: src-tauri/src/core/gen.rs
//! AI element generation — native port of `src/main/gen.ts`. Turn a text
//! description into a usable PNG graphic.
//!
//! Providers (switchable; the *id* persists in a small non-secret JSON, the
//! BFL key lives in env or the OS keychain behind `core/secrets.rs`):
//!   - `pollinations` (default) — free, no key. GET image.pollinations.ai.
//!   - `bfl` — Black Forest Labs FLUX (paid): POST job, poll, download.
//!
//! Generated images are normalized to PNG and saved into the `assets` table
//! (as a data URI, Electron parity) plus a `cache/gen/<id>.png` file, so they
//! appear in the editor's "your graphics" library automatically.

use std::fs;
use std::io::Cursor;
use std::path::PathBuf;
use std::sync::OnceLock;
use std::time::Duration;

use rusqlite::Connection;
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::core::db::now;
use crate::core::secrets;

const POLLINATIONS_ENDPOINT: &str = "https://image.pollinations.ai/prompt";
const BFL_ENDPOINT: &str = "https://api.bfl.ai/v1/flux-pro-1.1";
const BFL_RESULT_ENDPOINT: &str = "https://api.bfl.ai/v1/get_result";
/// FLUX job polling ceiling (Electron `MAX_POLL_SECONDS` parity).
const MAX_POLL_SECS: u64 = 120;

pub const GEN_PROVIDERS: [&str; 2] = ["pollinations", "bfl"];

fn http() -> &'static reqwest::Client {
    static CLIENT: OnceLock<reqwest::Client> = OnceLock::new();
    CLIENT.get_or_init(|| {
        // Generation can be slow — a generous cap beats Electron's unbounded
        // fetch while still failing eventually instead of hanging forever.
        reqwest::Client::builder()
            .timeout(Duration::from_secs(240))
            .build()
            .expect("reqwest client")
    })
}

/* ---------- DTOs (mirror src/shared/api.ts `gen.*`) ---------- */

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GenAsset {
    pub id: String,
    pub name: String,
    pub kind: String,
    pub data_uri: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GenResult {
    pub ok: bool,
    pub asset: Option<GenAsset>,
    pub error: Option<String>,
}

fn ok_result(id: String, name: String, data_uri: String) -> GenResult {
    GenResult {
        ok: true,
        asset: Some(GenAsset { id, name, kind: "png".into(), data_uri }),
        error: None,
    }
}

fn err_result(msg: impl Into<String>) -> GenResult {
    GenResult { ok: false, asset: None, error: Some(msg.into()) }
}

/* ---------- non-secret provider state ---------- */

fn gen_config_path(data_dir: &std::path::Path) -> PathBuf {
    data_dir.join("gen-config.json")
}

/// Active provider id — the file mirrors Electron's `gen-config.json` name but
/// stores only `{ "provider" }` (never the BFL key).
pub fn provider(data_dir: &std::path::Path) -> String {
    let cfg: Value = fs::read_to_string(gen_config_path(data_dir))
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or(Value::Null);
    match cfg.get("provider").and_then(|v| v.as_str()) {
        Some("bfl") => "bfl".to_string(),
        _ => "pollinations".to_string(),
    }
}

pub fn set_provider(data_dir: &std::path::Path, p: &str) -> bool {
    if !GEN_PROVIDERS.contains(&p) {
        return false;
    }
    let cfg = serde_json::json!({ "provider": p });
    let write = || -> std::io::Result<()> {
        let path = gen_config_path(data_dir);
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::write(path, serde_json::to_vec(&cfg).unwrap_or_default())
    };
    write().is_ok()
}

/// Save the BFL key (keychain only) and switch to BFL (Electron parity).
pub fn set_api_key(data_dir: &std::path::Path, key: &str) -> bool {
    let key = key.trim();
    if key.is_empty() {
        return false;
    }
    if !secrets::write_secret("gen/bfl", key) {
        return false;
    }
    set_provider(data_dir, "bfl")
}

/// True when the active provider can generate right now: pollinations is
/// keyless; bfl needs `BFL_API_KEY` (env) or the OS keychain.
pub fn configured(data_dir: &std::path::Path) -> bool {
    if provider(data_dir) == "pollinations" {
        return true;
    }
    secrets::flux_api_key().is_some()
}

/* ---------- HTTP ---------- */

/// Pollinations GET URL (parity with the pure `pollinationsUrl` helper).
fn pollinations_url(prompt: &str, opts: &Option<GenOpts>) -> String {
    let w = opts.as_ref().and_then(|o| o.width).unwrap_or(768);
    let h = opts.as_ref().and_then(|o| o.height).unwrap_or(768);
    let seed = opts
        .as_ref()
        .and_then(|o| o.seed)
        .unwrap_or_else(|| fastrand_ish() as u32);
    let encoded = urlencoding::encode(prompt);
    format!(
        "{POLLINATIONS_ENDPOINT}/{encoded}?width={w}&height={h}&nologo=true&seed={seed}"
    )
}

/// Cheap non-crypto random seed (Electron used `Math.random()`; any value is
/// fine — only provider-side determinism matters and that is per-seed).
fn fastrand_ish() -> u64 {
    let nanos = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.subsec_nanos() as u64)
        .unwrap_or(0);
    let pid = std::process::id() as u64;
    (nanos ^ (pid << 32)) & 0xFFFF_FFFF
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GenOpts {
    pub width: Option<u32>,
    pub height: Option<u32>,
    #[serde(default)]
    pub seed: Option<u32>,
}

/// Sniff a buffer's image type from its magic bytes (`sniffImageType` parity).
fn sniff(buf: &[u8]) -> &'static str {
    if buf.len() > 8 && buf[0] == 0x89 && buf[1] == 0x50 && buf[2] == 0x4e && buf[3] == 0x47 {
        "png"
    } else if buf.len() > 3 && buf[0] == 0xff && buf[1] == 0xd8 && buf[2] == 0xff {
        "jpeg"
    } else if buf.len() > 12 && &buf[0..4] == b"RIFF" && &buf[8..12] == b"WEBP" {
        "webp"
    } else {
        "png"
    }
}

/// Normalize to PNG bytes (Electron rasterized non-PNG provider output with
/// sharp; the `image` crate covers JPEG + WebP decode here).
fn to_png(buf: &[u8]) -> Result<Vec<u8>, String> {
    if sniff(buf) == "png" {
        return Ok(buf.to_vec());
    }
    let img = image::load_from_memory(buf).map_err(|e| {
        format!("provider returned {} that the native decoder could not convert — retry (PNG/JPEG responses convert fine): {e}", sniff(buf))
    })?;
    let mut out = Cursor::new(Vec::new());
    img.write_to(&mut out, image::ImageFormat::Png)
        .map_err(|e| format!("PNG encode failed: {e}"))?;
    Ok(out.into_inner())
}

/// Poll a BFL job until Ready and return the sample URL (`pollBflResult` parity).
async fn poll_bfl_result(id: &str, api_key: &str) -> Result<String, String> {
    let started = std::time::Instant::now();
    loop {
        let res = http()
            .get(BFL_RESULT_ENDPOINT)
            .query(&[("id", id)])
            .header("Authorization", format!("Bearer {api_key}"))
            .header("Accept", "application/json")
            .send()
            .await
            .map_err(|e| format!("FLUX status check failed — {e}"))?;
        if !res.status().is_success() {
            return Err(format!("FLUX status check failed ({})", res.status().as_u16()));
        }
        let data: Value = res.json().await.map_err(|e| format!("FLUX status parse — {e}"))?;
        if let Some(err) = data.get("error").and_then(|v| v.as_str()) {
            return Err(err.to_string());
        }
        if data.get("status").and_then(|v| v.as_str()) == Some("Ready") {
            if let Some(sample) = data
                .get("result")
                .and_then(|v| v.get("sample"))
                .and_then(|v| v.as_str())
            {
                return Ok(sample.to_string());
            }
        }
        if started.elapsed().as_secs() > MAX_POLL_SECS {
            return Err("FLUX generation timed out".into());
        }
        // Blocking sleep on a worker thread — keeps the poll loop async while
        // avoiding a direct tokio dependency for one timer.
        let _ = tauri::async_runtime::spawn_blocking(move || std::thread::sleep(Duration::from_millis(1500))).await;
    }
}

/// Fetch image bytes from the active provider (no DB access — callers insert
/// the asset row themselves so the rusqlite connection never crosses `.await`).
async fn fetch_bytes(
    data_dir: &std::path::Path,
    prompt: &str,
    opts: &Option<GenOpts>,
) -> Result<Vec<u8>, String> {
    let provider = provider(data_dir);
    if provider == "pollinations" {
        let res = http()
            .get(pollinations_url(prompt, opts))
            .send()
            .await
            .map_err(|e| format!("Pollinations returned an error — {e}"))?;
        if !res.status().is_success() {
            return Err(format!("Pollinations returned {}", res.status().as_u16()));
        }
        return Ok(res.bytes().await.map_err(|e| e.to_string())?.to_vec());
    }
    // bfl / FLUX
    let key = secrets::flux_api_key().ok_or("Add a Black Forest Labs API key to use FLUX.")?;
    let w = opts.as_ref().and_then(|o| o.width).unwrap_or(1024);
    let h = opts.as_ref().and_then(|o| o.height).unwrap_or(1024);
    let job_res = http()
        .post(BFL_ENDPOINT)
        .header("Authorization", format!("Bearer {key}"))
        .header("Content-Type", "application/json")
        .body(
            serde_json::json!({ "prompt": prompt.trim(), "width": w, "height": h })
                .to_string(),
        )
        .send()
        .await
        .map_err(|e| format!("FLUX request failed — {e}"))?;
    if !job_res.status().is_success() {
        return Err(format!("FLUX request failed ({})", job_res.status().as_u16()));
    }
    let job: Value = job_res.json().await.map_err(|e| format!("FLUX job parse — {e}"))?;
    let id = job
        .get("id")
        .and_then(|v| v.as_str())
        .ok_or("FLUX returned no job id")?;
    let sample = poll_bfl_result(id, &key).await?;
    let img_res = http()
        .get(&sample)
        .send()
        .await
        .map_err(|e| format!("FLUX image download failed — {e}"))?;
    if !img_res.status().is_success() {
        return Err(format!("FLUX image download failed ({})", img_res.status().as_u16()));
    }
    Ok(img_res.bytes().await.map_err(|e| e.to_string())?.to_vec())
}

/// Fetch + normalize a generated image (no DB access — the rusqlite
/// connection lives behind a `with_db` sync scope and must never cross
/// `.await`; commands call this async, then hand the PNG to `persist_asset`
/// inside a short DB block). Returns the error string directly so callers can
/// build a `GenResult`.
pub async fn fetch_generated(
    data_dir: &std::path::Path,
    prompt: &str,
    opts: &Option<GenOpts>,
) -> Result<Vec<u8>, String> {
    let clean = prompt.trim();
    if clean.is_empty() {
        return Err("Enter a description first.".into());
    }
    let bytes = fetch_bytes(data_dir, clean, opts).await?;
    if bytes.is_empty() {
        return Err("Empty image received from provider.".into());
    }
    to_png(&bytes)
}

/// Write `<id>.png` into `cache/gen/` + insert the `assets` row (data URI).
/// Mirrors Electron's `GenService.generate` tail exactly.
pub fn persist_asset(
    conn: &Connection,
    cache_dir: &std::path::Path,
    name: &str,
    png: &[u8],
) -> GenResult {
    use base64::Engine as _;
    let id = {
        let millis = chrono::Utc::now().timestamp_millis();
        let salt = &uuid::Uuid::new_v4().simple().to_string()[..6];
        format!("gen-{millis}-{salt}")
    };
    let gen_dir = cache_dir.join("gen");
    if let Err(e) = fs::create_dir_all(&gen_dir) {
        return err_result(format!("gen cache dir: {e}"));
    }
    if let Err(e) = fs::write(gen_dir.join(format!("{id}.png")), png) {
        return err_result(format!("gen cache write: {e}"));
    }
    let data_uri = format!(
        "data:image/png;base64,{}",
        base64::engine::general_purpose::STANDARD.encode(png)
    );
    let label = name.chars().take(60).collect::<String>();
    if let Err(e) = conn.execute(
        "INSERT INTO assets (id, name, kind, data, created_at) VALUES (?1, ?2, ?3, ?4, ?5)",
        rusqlite::params![id, label, "png", data_uri, now()],
    ) {
        return err_result(format!("asset write: {e}"));
    }
    ok_result(id, label, data_uri)
}
